src/tunnel.cc

```diff
@@ -638,10 +638,10 @@ TunnelStateData::readClient(char *, size_t len, Comm::Flag errcode, int xerrno)
     if (errcode == Comm::ERR_CLOSING)
         return;
 
-    if (len > 0) {
+    if (len > 0)
         client.bytesIn(len);
-        statCounter.client_http.kbytes_in += len;
-    }
+
+    ReadFromClient(al, len, errcode);
```

Comment:
As we discussed, `ReadFromClient()` and other three functions minimize duplication by collecting global statistics, but do not collect local (transaction-related) statistics as well. For example, in this context the local statistics is calculated by the `client.bytesIn(len)` call. In `HttpStateData::readReply()` it would be probably `HttpStateData::payloadSeen` variable. We have also ClientHttpRequest::out::size calculated in both HTTP and FTP code. Some of these context-related variables are used for other format codes, e.g., `"<st"` is gathered from `al->http.clientReplySz`, that in turn is calculated from `ClientHttpRequest::out`.
In general, all these transaction-related bits seem rather unsystematic: I do not see an (easy) way how we can neatly adjust further, moving them into `ReadFromClient()` (and similar) functions.